ReSharper.FSharp/src/FSharp.Psi.Services/src/Generate/GenerateProvider.fs

```diff
@@ -195,29 +196,91 @@ type FSharpOverridingMembersBuilder() =
 
-        let currentIndent = typeRepr.Indent
-        let desiredIndent = typeDecl.Indent + typeDecl.GetIndentSize()
+        let indentSize = typeDecl.GetIndentSize()
+        let desiredIndent = typeDecl.Indent + indentSize
 
         addNodesBefore typeRepr.FirstChild [
             NewLine(typeRepr.GetLineEnding())
-            Whitespace(currentIndent)
+            Whitespace(desiredIndent)
         ] |> ignore
 
-        // to have good indents for begin/end keywords which might start out on different indents
-        shiftNode (-currentIndent) typeRepr
-        shiftNode desiredIndent typeRepr
+        let normalizeReprEnd (beginToken: ITokenNode) (endToken: ITokenNode) =
+            if isNull beginToken || isNull endToken then ()
+            elif beginToken.Indent > endToken.Indent then
+                let diff = beginToken.Indent - endToken.Indent
+                addNodeBefore endToken (Whitespace(diff))
+            elif endToken.Indent > beginToken.Indent && isFirstMeaningfulNodeOnLine endToken then
+                let diff = endToken.Indent - beginToken.Indent
+                shiftWithWhitespaceBefore -diff endToken
+
+        let beginToken, endToken =
+            match typeRepr with
+            | :? IObjectModelTypeRepresentation as objRepr ->
+                normalizeReprEnd objRepr.BeginKeyword objRepr.EndKeyword
+                objRepr.BeginKeyword, objRepr.EndKeyword
+            | :? IRecordRepresentation as recordRepr ->
+                normalizeReprEnd recordRepr.LeftBrace recordRepr.RightBrace
+                let fieldDeclarations = recordRepr.FieldDeclarations 
+                if fieldDeclarations.Count > 1 then
+                    fieldDeclarations
+                    |> Seq.indexed
+                    |> Seq.skip 1
+                    |> Seq.iter (fun (idx, f) ->
+                        let prevField = fieldDeclarations[idx - 1]
+                        if f.StartLine <> prevField.StartLine && f.Indent > fieldDeclarations[0].Indent then
+                            let diff = f.Indent - fieldDeclarations[0].Indent
+                            shiftWithWhitespaceBefore -diff f)
+                recordRepr.LeftBrace, recordRepr.RightBrace
+            | _ -> null, null
+
+        let rec deindentEmptyLines indentationTarget (nodes: seq<ITreeNode>) =
+            nodes
+            |> Seq.iter
+                (fun n ->
+                    if isAtEmptyLine n then
+                        let nextToken =  n.GetNextToken()
+                        if isNotNull nextToken then
+                            let diff = nextToken.Indent - indentationTarget
+                            match n with
+                            | :? Whitespace as whitespace -> if diff > 0 then shiftWhitespaceBefore -diff whitespace
+                            | _ -> ()
+                    else
+                        deindentEmptyLines indentationTarget (n.Children()))
```

Comment:
@dawedawe It looks like a wrong place to do it, and it adds a lot of complication to the code. Have you considered updating `shiftNode` or `shiftWithWhitespaceBefore`, so there's no need to process some of the tokens separately in every feature? It seems like it should be a ~two line fix somewhere in there instead of ~twenty+ lines of special checks here.

Reply:
I thought it wouldn't be acceptable to change such an infrastructure code, so I shied away from it.
But if you're okay with it, all the better. There's no impact on other existing tests.

To let existing shifting functions work on the representation nodes so that whitespace in empty lines is also shifted, I still had to rearrange some things...
I think it works now as you want for representations, but not for empty lines which are just in the declaration.
Given that many people use `trim_trailing_whitespace` in their editorconfig, I'm not sure how much effort we should put into it. But, of course, it's your decision.

Reply:
> I thought it wouldn't be acceptable to change such an infrastructure code, so I shied away from it.
> But if you're okay with it, all the better. There's no impact on other existing tests.

It's more than welcome to improve it. 🙂